### DISCUSSION TITLE
Question about how to use tsk_treeseq_genealogical_nearest_neighbours function of C API

### DISCUSSION THREAD

**Discussion by T-Funayama** [tskit-dev/tskit#2787] [Q&A]
Hi

I am trying to load trees file and run tsk_treeseq_genealogical_nearest_neighbours function of C API. Could you tell me argument data format of tsk_treeseq_genealogical_nearest_neighbours function,  how to use C API and so on. If possible, Could you please send me the example code?

I look forward to hearing from you soon.
Thank you.

**Comment by petrelharp** [comment]
Well, it seems that function is not documented in the C API, which means it might be subject to change, but, here's the reference:
  https://github.com/tskit-dev/tskit/blob/main/c/tskit/trees.h#L908C17-L908C17

> how to use C API and so on

For this, I'd start with https://tskit.dev/tskit/docs/latest/c-api.html and post specific questions if you run into difficulties?

**Reply by T-Funayama** [reply]
Thank you for your reply. 
I have learned by reading the manual. 

I don't understand reference_sets and reference_set_size of argument data in tsk_treeseq_genealogical_nearest_neighbours function.
Could you tell me how to make data which are reference_sets and reference_set_size? Also, I sent to you a script that I am in the process of creating. It would be happy if you could advised about its script.

```c
#include <stdio.h>
#include <stdlib.h>
#include <tskit.h>

int
main(int argc, char **argv)
{
    int ret, gnn;
    tsk_treeseq_t ts;
    const tsk_id_t *samples, *reference_sets;
    tsk_size_t samples_size, reference_set_size;


    ret = tsk_treeseq_load(&ts, "1kg_chr20.trees",0);
    if (ret < 0) {
        /* Error condition. Free and exit */
        tsk_treeseq_free(&ts);
        fprintf(stderr, "%s", tsk_strerror(ret));
        exit(EXIT_FAILURE);
    }
    

    samples = tsk_treeseq_get_samples(&ts);
    samples_size = tsk_treeseq_get_num_samples(&ts);
    
    /*  */
    reference_sets = 
    reference_set_size = 

    gnn = tsk_treeseq_genealogical_nearest_neighbours(&ts, samples, samples_size, reference_sets, reference_set_size);

    return EXIT_SUCCESS;
}
```

**Reply by jeromekelleher** [reply]
Hi @T-Funayama, welcome :wave:

The reference_sets and reference_set_sizes correspond to the ``sample_sets`` argument in the [python API](https://tskit.dev/tskit/docs/stable/python-api.html#tskit.TreeSequence.genealogical_nearest_neighbours).

Before we go any further, though, is there a particular reason that you don't want to use the Python interface here?

**Reply by T-Funayama** [reply]
Thank you so much.
I usually use the Python interface. I am using C API as next step of learning.

**Reply by jeromekelleher** [reply]
I see, that's great you're interested in learning the C API! I'd suggest sticking to the documented API then- there's no real point in trying to figure out how to use stuff we've deliberately hidden away.
